silence/silence.go

```diff
@@ -0,0 +1,792 @@
+	}
+	// Validate modification based on current silence state.
+	switch getState(sil, now) {
+	case StateActive:
+		if *start != *sil.StartsAt {
+			return nil, fmt.Errorf("start time of active silence cannot be modified")
+		}
+		if protoBefore(end, now) {
+			return nil, fmt.Errorf("end time cannot be set into the past")
+		}
+	case StatePending:
+		if protoBefore(start, now) {
+			return nil, fmt.Errorf("start time cannot be set into the past")
+		}
+	case StateExpired:
+		return nil, fmt.Errorf("expired silence must not be modified")
+	default:
+		panic("unknown silence state")
+	}
+
+	sil = cloneSilence(sil)
+	sil.StartsAt = start
+	sil.EndsAt = end
+	sil.UpdatedAt = now
+
+	return sil, nil
+}
+
+// AddComment adds a new comment to the silence with the given ID.
+func (s *Silences) AddComment(id string, author, comment string) error {
+	panic("not implemented")
+}
+
+// QueryParam expresses parameters along which silences are queried.
+type QueryParam func(*query) error
+
+type query struct {
+	ids     []string
+	filters []silenceFilter
+}
+
+// silenceFilter is a function that returns true if a silence
+// should be dropped from a result set for a given time.
+type silenceFilter func(*pb.Silence, *timestamp.Timestamp) (bool, error)
+
+var errNotSupported = errors.New("query parameter not supported")
+
+// QIDs configures a query to select the given silence IDs.
+func QIDs(ids ...string) QueryParam {
+	return func(q *query) error {
+		q.ids = append(q.ids, ids...)
+		return nil
+	}
+}
+
+// QTimeRange configures a query to search for silences that are active
+// in the given time range.
+// TODO(fabxc): not supported yet.
+func QTimeRange(start, end time.Time) QueryParam {
+	return func(q *query) error {
+		return errNotSupported
+	}
+}
+
+// QMatches returns silences that match the given label set.
+func QMatches(set map[string]string) QueryParam {
+	return func(q *query) error {
+		f := func(s *pb.Silence, _ *timestamp.Timestamp) (bool, error) {
+			// TODO(fabxc): we compile every regexp matcher of a silence
+			// each time we check it against a label set.
+			// This could be notably slower than the old caching behavior.
+			// With the protobuf type not being extensible, we need a more
+			// efficient solution without wrapping the silence in another layer.
+			for _, m := range s.Matchers {
+				switch m.Type {
+				case pb.Matcher_EQUAL:
+					if set[m.Name] != m.Pattern {
+						return false, nil
+					}
+				case pb.Matcher_REGEXP:
+					re, err := regexp.Compile(m.Pattern)
+					if err != nil {
+						return false, err
+					}
+					if !re.MatchString(set[m.Name]) {
+						return false, nil
+					}
+				}
+			}
+			// All matchers applied to the given set and the silence
+			// passes as a result.
+			return true, nil
+		}
+		q.filters = append(q.filters, f)
+		return nil
+	}
+}
+
+// SilenceState describes the state of a silence based on its time range.
+type SilenceState string
```

Comment:
There'll possibly be `State(sil) SilenceState` function at some point. This would suffer from the same problem but I'd imagine `st := silence.State(sil)` to be more frequent than `var st silence.SilenceState; st = silence.State(sil)`. The latter in return would be overly verbose the other way around as well `var st silence.State; st = silence.SilenceState(sil)`.

`GetState` doesn't seem like an idiomatic replacement either. Right now there's little reason to ever refernce the type in importing code.